### PR TITLE
Fix the handling of duplicate props

### LINKED ISSUE #7969: Crash if field is defined twice from duplicate parameter
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0A%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%20%20sig%20%7B%20params%28x%3A%20Integer%2C%20x%3A%20String%29.void%20%7D%0A%20%20def%20initialize%28x%3A%2C%20x%3A%20nil%29%0A%20%20%20%20%40x%20%3D%20T.let%28x%2C%20T.nilable%28String%29%29%0A%20%20%20%20%40x%20%3D%20T.let%28x%2C%20T.nilable%28String%29%29%0A%20%20end%0Aend%0A)

```ruby
# typed: strict

class A
  extend T::Sig
  sig { params(x: Integer, x: String).void }
  def initialize(x:, x: nil)
    @x = T.let(x, T.nilable(String))
    @x = T.let(x, T.nilable(String))
  end
end

```

#### Observed output

```
msg="Bad core::Types::equiv in resolveField" path="-e" priorField="<no-field-or-static-field>" resultTypeNull="true" castTypeNull="false"
source="# typed: strict\n\nclass A\n  extend T::Sig\n  sig { params(x: Integer, x: String).void }\n  def initialize(x:, x: nil)\n    @x = T.let(x, T.nilable(String))\n    @x = T.let(x, T.nilable(String))\n  end\nend\n\n"
Exception::raise(): isSubTypeUnderConstraint(, ClassType): unreachable

???: Exception resolving (backtrace is above) https://srb.help/1001

editor.rb:6: duplicate argument name x https://srb.help/2001
     6 |  def initialize(x:, x: nil)
                             ^^^^^^

editor.rb:5: Hash key x is duplicated https://srb.help/3011
     5 |  sig { params(x: Integer, x: String).void }
                                   ^
    editor.rb:5: First occurrence of x hash key
     5 |  sig { params(x: Integer, x: String).void }
                       ^
Errors: 3
```

#### Expected behavior

Sorbet should not crash.

- - -

While this example seems contrived, it comes up somewhat naturally if you accidentally define a prop twice in a `T::Struct`:

```ruby
# typed: strict

class A < T::Struct
  const :foo, Integer
  const :foo, T.nilable(String)
end
```

### PR DESCRIPTION
**Pull request by elliottt**
The previous handling of duplicate args relied on fixing the error during parsing, however the rewrite pass can also introduce new methods, and they don't benefit from that fix. This PR resolves this issue by also mangling duplicate fields in the local_vars pass, which is the last pass of the indexing phase.

An additional consequence of resolving these name issues in the local_vars pass is that the name table now needs to be unfrozen while running that pass.

### Motivation
Fixes #7969

### Test plan

See included automated tests.

### CODE REVIEW

File: local_vars/local_vars.cc [hunk 1]
```diff
@@ -47,16 +47,29 @@ class LocalNameInserter {
...
+            // Mangle any duplicate keyword arguments.
+            if (isKeyword) {
+                auto it = absl::c_find_if(seen, [name](auto &existing) { return existing.name == name; });
+                if (it != seen.end()) {
+                    return ctx.state.freshNameUnique(core::UniqueNameKind::MangledKeywordArg, name, pos);
+                }
+            }
+
+            return name;
+        };
```
Comment (jez) [inline local_vars/local_vars.cc:64]: Should we make this a helper function, not a lambda? I don't think we should need to be allocating a closure every time this function gets called.

File: main/pipeline/pipeline.cc [hunk 2]
```diff
@@ -155,8 +155,9 @@ ast::ExpressionPtr runRewriter(core::GlobalState &gs, core::FileRef file, ast::E
 }
 
 ast::ParsedFile runLocalVars(core::GlobalState &gs, ast::ParsedFile tree) {
-    Timer timeit(gs.tracer(), "runLocalVars", {{"file", string(tree.file.data(gs).path())}});
     core::MutableContext ctx(gs, core::Symbols::root(), tree.file);
+    Timer timeit(gs.tracer(), "runLocalVars", {{"file", string(tree.file.data(gs).path())}});
```
Comment (jez) [inline main/pipeline/pipeline.cc:159]: Any reason why not to have this at the top of the function still?

Comment (elliottt) [reply main/pipeline/pipeline.cc:159]: I was making it consistent with the structure of `runRewriter`. Happy to move it back, as it's hopefully a no-op :)

File: local_vars/local_vars.cc [hunk 3]
```diff
@@ -45,18 +45,33 @@ class LocalNameInserter {
...
     };
 
+    // Handle the mangling of keyword argument names, if the name passed has already been seen in the argument list.
+    core::NameRef mangleKeyword(core::MutableContext ctx, const vector<NamedArg> &seen, bool isKeyword,
+                                core::NameRef name, uint32_t pos) const {
+        // Mangle any duplicate keyword arguments.
+        if (isKeyword) {
+            auto it = absl::c_find_if(seen, [name](auto &existing) { return existing.name == name; });
+            if (it != seen.end()) {
+                return ctx.state.freshNameUnique(core::UniqueNameKind::MangledKeywordArg, name, pos);
```
Comment (jez) [inline local_vars/local_vars.cc:55]: The place where we mangle the names we should be reporting the error. This is how it works for e.g. mangled methods.

Mangling is a mechanism to recover from errors, and we need to ensure that that error is reported to the user, or else they'll get weird behavior and not realize what the source of it was.

Before we started making these changes, namer reported the error and namer mangled the keywords. Then, the parser did both. But now, the parser reports the error while local_vars does the mangling. That means that for any violations which are introduced between the parser and the local_vars phase (e.g. any violations in rewriter passes), we'll have missing errors.

**Review by jez — CHANGES_REQUESTED** [review]
^ Mostly looks good but I think we need to make one change

File: main/pipeline/pipeline.cc (hunk 2 above)
Comment (jez) [reply main/pipeline/pipeline.cc:159]: No preference.

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_R813tEWeR7rZPo
→ https://go/builds/bui_R813QYQtcqEN4h
→ https://go/builds/bui_R813F4G7Dv2ErK

**Review by jez — APPROVED** [review]

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_R82cmdqVvhtcdg
→ https://go/builds/bui_R82ceuJCg94zzu
→ https://go/builds/bui_R82cpiM88eCg4a

**Pull request merged**